File: python/src/fantasy_football_prediction_model/projections/scoring.py
"""Configurable fantasy scoring from projected football statistics."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Mapping

from fantasy_football_prediction_model.config import ScoringConfig, ScoringRules


@dataclass(frozen=True, slots=True)
class ScoreBreakdown:
    """Transparent arithmetic for one scoring result."""

    total: float
    passing: float
    rushing: float
    receiving: float
    misc: float


def score_stats(stats: Mapping[str, float | None], rules: ScoringRules) -> ScoreBreakdown:
    """Score a mapping of snake_case or camelCase projected statistics."""
    def g(*keys: str) -> float:
        for key in keys:
            value = stats.get(key)
            if value is not None:
                return float(value)
        return 0.0

    passing_yards = g("passing_yards", "passingYards")
    passing_tds = g("passing_tds", "passing_touchdowns", "passingTouchdowns")
    interceptions = g("interceptions")
    rushing_yards = g("rushing_yards", "rushingYards")
    rushing_tds = g("rushing_tds", "rushing_touchdowns", "rushingTouchdowns")
    receptions = g("receptions")
    receiving_yards = g("receiving_yards", "receivingYards")
    receiving_tds = g("receiving_tds", "receiving_touchdowns", "receivingTouchdowns")
    fumbles = g("fumbles_lost", "fumblesLost")

    passing = 0.0
    if rules.passing.yards_per_point:
        passing += passing_yards / rules.passing.yards_per_point
    passing += passing_tds * rules.passing.touchdown
    passing += interceptions * rules.passing.interception

    rushing = 0.0
    if rules.rushing.yards_per_point:
        rushing += rushing_yards / rules.rushing.yards_per_point
    rushing += rushing_tds * rules.rushing.touchdown

    receiving = receptions * rules.receiving.reception
    if rules.receiving.yards_per_point:
        receiving += receiving_yards / rules.receiving.yards_per_point
    receiving += receiving_tds * rules.receiving.touchdown

    misc = fumbles * rules.misc.fumble_lost
    total = passing + rushing + receiving + misc
    return ScoreBreakdown(
        total=float(total),
        passing=float(passing),
        rushing=float(rushing),
        receiving=float(receiving),
        misc=float(misc),
    )


def score_total(stats: Mapping[str, float | None], rules: ScoringRules) -> float:
    return score_stats(stats, rules).total


def rules_from_preset(scoring: ScoringConfig, preset: str | None = None) -> ScoringRules:
    name = preset or scoring.default_preset
    if name not in scoring.presets:
        raise KeyError(f"Unknown scoring preset '{name}'. Known: {sorted(scoring.presets)}")
    return scoring.presets[name].rules


def rules_to_export_dict(rules: ScoringRules) -> dict[str, float]:
    return {
        "passing_yards_per_point": rules.passing.yards_per_point,
        "passing_touchdown": rules.passing.touchdown,
        "interception": rules.passing.interception,
        "passing_two_point": rules.passing.two_point_conversion,
        "rushing_yards_per_point": rules.rushing.yards_per_point,
        "rushing_touchdown": rules.rushing.touchdown,
        "rushing_two_point": rules.rushing.two_point_conversion,
        "reception": rules.receiving.reception,
        "receiving_yards_per_point": rules.receiving.yards_per_point,
        "receiving_touchdown": rules.receiving.touchdown,
        "receiving_two_point": rules.receiving.two_point_conversion,
        "fumble_lost": rules.misc.fumble_lost,
    }


def custom_rules(base: ScoringRules, overrides: dict[str, Any]) -> ScoringRules:
    """Return a copy of ``base`` with selected numeric overrides applied."""
    payload = {
        "passing": base.passing.model_dump(),
        "rushing": base.rushing.model_dump(),
        "receiving": base.receiving.model_dump(),
        "misc": base.misc.model_dump(),
    }
    mapping = {
        "reception": ("receiving", "reception"),
        "passing_yards_per_point": ("passing", "yards_per_point"),
        "passing_touchdown": ("passing", "touchdown"),
        "interception": ("passing", "interception"),
        "rushing_yards_per_point": ("rushing", "yards_per_point"),
        "receiving_yards_per_point": ("receiving", "yards_per_point"),
        "rushing_touchdown": ("rushing", "touchdown"),
        "receiving_touchdown": ("receiving", "touchdown"),
        "fumble_lost": ("misc", "fumble_lost"),
    }
    for key, value in overrides.items():
        if key in mapping:
            group, field = mapping[key]
            payload[group][field] = value
    return ScoringRules.model_validate(payload)

File: web/src/lib/scoring.ts
export type ScoringRules = {
  passingYardsPerPoint: number;
  passingTouchdown: number;
  interception: number;
  rushingYardsPerPoint: number;
  rushingTouchdown: number;
  reception: number;
  receivingYardsPerPoint: number;
  receivingTouchdown: number;
  fumbleLost: number;
};

export const PRESETS: Record<string, ScoringRules> = {
  standard: {
    passingYardsPerPoint: 25,
    passingTouchdown: 4,
    interception: -2,
    rushingYardsPerPoint: 10,
    rushingTouchdown: 6,
    reception: 0,
    receivingYardsPerPoint: 10,
    receivingTouchdown: 6,
    fumbleLost: -2,
  },
  half_ppr: {
    passingYardsPerPoint: 25,
    passingTouchdown: 4,
    interception: -2,
    rushingYardsPerPoint: 10,
    rushingTouchdown: 6,
    reception: 0.5,
    receivingYardsPerPoint: 10,
    receivingTouchdown: 6,
    fumbleLost: -2,
  },
  ppr: {
    passingYardsPerPoint: 25,
    passingTouchdown: 4,
    interception: -2,
    rushingYardsPerPoint: 10,
    rushingTouchdown: 6,
    reception: 1,
    receivingYardsPerPoint: 10,
    receivingTouchdown: 6,
    fumbleLost: -2,
  },
};

export type StatLine = {
  passingYards?: number | null;
  passingTouchdowns?: number | null;
  interceptions?: number | null;
  rushingYards?: number | null;
  rushingTouchdowns?: number | null;
  receptions?: number | null;
  receivingYards?: number | null;
  receivingTouchdowns?: number | null;
  fumblesLost?: number | null;
};

export function scoreStats(stats: StatLine, rules: ScoringRules): number {
  const passingYards = stats.passingYards ?? 0;
  const passingTds = stats.passingTouchdowns ?? 0;
  const interceptions = stats.interceptions ?? 0;
  const rushingYards = stats.rushingYards ?? 0;
  const rushingTds = stats.rushingTouchdowns ?? 0;
  const receptions = stats.receptions ?? 0;
  const receivingYards = stats.receivingYards ?? 0;
  const receivingTds = stats.receivingTouchdowns ?? 0;
  const fumbles = stats.fumblesLost ?? 0;

  let total = 0;
  if (rules.passingYardsPerPoint) total += passingYards / rules.passingYardsPerPoint;
  total += passingTds * rules.passingTouchdown;
  total += interceptions * rules.interception;
  if (rules.rushingYardsPerPoint) total += rushingYards / rules.rushingYardsPerPoint;
  total += rushingTds * rules.rushingTouchdown;
  total += receptions * rules.reception;
  if (rules.receivingYardsPerPoint) total += receivingYards / rules.receivingYardsPerPoint;
  total += receivingTds * rules.receivingTouchdown;
  total += fumbles * rules.fumbleLost;
  return total;
}

export function getPreset(name: string): ScoringRules {
  return PRESETS[name] ?? PRESETS.ppr;
}
